keps/sig-network/4427-relaxed-dns-search-validation/README.md

```diff
@@ -0,0 +1,621 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently, Kubernetes validates search string in the `dnsConfig.searches` according to [RFC-1123](https://datatracker.ietf.org/doc/html/rfc1123)
+which defines restrictions for hostnames. While most DNS names identify hosts, there are record types (like SRV) that don't. For these, it's less clear
+whether hostname restritions apply, for example [RFC-1035 Section 2.3.1](https://datatracker.ietf.org/doc/html/rfc1035#section-2.3.1) points out
+that it's better to stick with valid host names but also states that labels must meet the hostname requirements.
+
+In practice, legcay workloads sometimes include an underscore (`_`) in DNS names and DNS servers will generally allow this.
+
+This KEP proposes relaxing the checks on DNS search strings only. Allowing these values in the `searches` field of `dnsConfig` allows pods to
+resolve short names properly in cases where the search string contains and underscore.
+
+## Motivation
+
+For workloads that resolve short DNS names where the full DNS name includes disallowed characters (like underscores),
+it’s not possible to configure search strings using dnsConfig. For example, if a pod needs to look up an SRV record `_sip._tcp.abc_d.example.com`
+using the short name of `_sip._tcp`, we would like to be able to add `abc_d.example.com` to the searches in the dnsConfig.
+
+Here’s an example configuration which would support this case:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: dns-example
+spec:
+  containers:
+    - name: test
+      image: nginx
+  dnsPolicy: "None"
+  dnsConfig:
+    nameservers:
+      - 1.2.3.4
+    searches:
+      - abc_d.example.com
+```
+
+However, this returns an error:
+
+```
+The Pod "dns-example" is invalid: spec.dnsConfig.searches[0]: Invalid value: "abc_d.example.com": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
+```
+
+Allowing underscores in the search string allows integration with legacy workloads without allowing anyone to define
+these names within Kubernetes. Since having underscores in a name creates other issues (such as inability to obtain a publicly trusted TLS certificate),
+search strings seem like the only area where this is likely to occur.
```

Comment:
Based on the updated text, the reason we want this specific change is not about "searching" vs "defining", it's about "hostnames in DNS" vs "non-hostnames in DNS". Theoretically, some future Kubernetes feature might result in CoreDNS creating SRV records based on Kubernetes API fields, and in that case, we might want to allow using underscores in those fields, right?

So the reason we want to change `searches` and we don't want to change anything else is because `searches` is (currently) the only API field whose value gets used for things related to non-hostname DNS data.

Reply:
Yes, i agree with that reasoning. If a future feature introduces that case, I could see having a discussion about whether allowing underscores there encourages good practices (the RFC does point out that sticking to the hostname rules leads to fewer problems, which matches my experience as well).